addons/website_sale/controllers/main.py

```diff
@@ -1862,6 +1862,9 @@ def shop_payment_transaction(self, order_id, access_token, **kwargs):
         if tools.float_compare(kwargs['amount'], order_sudo.amount_total, precision_rounding=order_sudo.currency_id.rounding):
             raise ValidationError(_("The cart has been updated. Please refresh the page."))
 
+        if not order_sudo.only_services and not order_sudo.carrier_id:
+            raise ValidationError(_("No shipping method is selected."))
```

Comment:
NB: could be integrated into `_check_cart_is_ready_to_be_paid` in the 17.0 fw-port ;)